packages/skill-lesson/hooks/use-progress.ts

```diff
@@ -64,45 +80,112 @@ export const useModuleProgress = ({module}: {module: SanityDocument}) => {
-    module: module.slug.current,
-  })
+  return {nextSection, moduleProgress}
+}
 
+export const getCompleted = ({
+  resources,
+  moduleProgress,
+  exercises,
+  nextExercise,
+}: {
+  resources: LessonResource[]
+  moduleProgress: LessonProgress[]
+  exercises: Exercise[]
+  nextExercise?: LessonResource | null
+}) => {
   const isCompleted = moduleProgress.length === resources.length
   const percentCompleted = Math.round(
     (moduleProgress.length / resources.length) * 100,
   )
 
-  if (
-    !isEmpty(
+  const hasProgress = !isEmpty(
+    find(moduleProgress, {
+      // if explainer
+      lessonSlug: get(find(exercises, {slug: nextExercise?.slug}), 'slug'),
+    }) ||
       find(moduleProgress, {
-        // if explainer
-        lessonSlug: get(find(exercises, {slug: nextExercise?.slug}), 'slug'),
-      }) ||
-        find(moduleProgress, {
-          lessonSlug: get(
-            find(exercises, {slug: nextExercise?.slug}),
-            'solution.slug',
-          ),
-        }),
-    )
-  ) {
-    setLastCompletedExercise(
-      nextSection.lessons[
-        indexOf(
-          nextSection.lessons,
-          find(nextSection.lessons, {slug: nextExercise?.slug}),
-        )
-      ],
-    )
-  }
+        lessonSlug: get(
+          find(exercises, {slug: nextExercise?.slug}),
+          'solution.slug',
+        ),
+      }),
+  )
+
+  const lastCompletedSolution = getLastCompletedSolution({
+    resources,
+    moduleProgress,
+  })
+
+  return {isCompleted, percentCompleted, lastCompletedSolution, hasProgress}
+}
+
+export const useModuleProgress = ({module}: {module: Module}) => {
+  const {sections} = module
+  const {data: userProgress = [], status: userProgressStatus} =
+    trpcSkillLessons.progress.get.useQuery()
+
+  const {exercises, resources} = extractExercisesAndResource(sections)
+
+  // storing this in state allows to refetch nextExercise
+  // in case the original next is already completed
+  const [lastCompletedExercise, setLastCompletedExercise] = React.useState(
+    first(exercises),
+  )
+
+  const {nextSection, moduleProgress} = getNextSectionAndModuleProgress({
+    userProgress,
+    resources,
+    lastCompletedExercise,
+    sections,
+    module,
+  })
+
+  const {data: nextExercise, status: nextExerciseStatus} =
+    trpcSkillLessons.lessons.getNextLesson.useQuery({
+      type: 'solution',
+      slug: lastCompletedExercise?.slug,
+      section: nextSection?.slug,
+      module: module.slug.current,
+    })
+
+  const {isCompleted, percentCompleted, lastCompletedSolution, hasProgress} =
+    getCompleted({
+      resources,
+      moduleProgress,
+      exercises,
+      nextExercise,
+    })
+
+  React.useEffect(() => {
+    if (hasProgress && nextSection) {
+      setLastCompletedExercise(
+        nextSection.lessons[
+          indexOf(
+            nextSection.lessons,
+            find(nextSection.lessons, {slug: nextExercise?.slug}),
+          )
+        ],
+      )
+    }
```

Comment:
this was NOT in an effect before and I think because it had logic and called a setter it should be. not super stoked on this.